VideoLocker/src/test/java/org/edx/mobile/test/http/OkHttpBaseTestCase.java

```diff
@@ -0,0 +1,206 @@
+        super.setUp();
+
+        server = new MockWebServer();
+        server.setDispatcher(new MockResponseDispatcher());
+        server.start();
+
+        context = RuntimeEnvironment.application;
+
+        // Set up a new config instance that serves the mock host url
+        JsonObject properties;
+        try {
+            InputStream in = context.getAssets().open("config/config.json");
+            JsonParser parser = new JsonParser();
+            JsonElement config = parser.parse(new InputStreamReader(in));
+            properties = config.getAsJsonObject();
+        } catch (Exception e) {
+            properties = new JsonObject();
+            logger.error(e);
+        }
+        properties.addProperty(API_HOST_URL, getBaseMockUrl());
+        Config.setInstance(new Config(properties));
+
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        server.shutdown();
+    }
+
+
+
+    /**
+     * Get the delay (in milliseconds) that should be used for delaying
+     * a successful network call response.
+     */
+    private static int calculateDelayForSuccess() {
+        float errorPercent = VARIANCE_PCT / 100f; // e.g., 20 / 100f == 0.2f
+        float lowerBound = 1f - errorPercent; // 0.2f --> 0.8f
+        float upperBound = 1f + errorPercent; // 0.2f --> 1.2f
+        float bound = upperBound - lowerBound; // 1.2f - 0.8f == 0.4f
+        float delayPercent = (random.nextFloat() * bound) +
+                lowerBound; // 0.8 + (rnd * 0.4)
+        return (int) (DELAY_MS * delayPercent);
+    }
+
+    /**
+     * Get the delay (in milliseconds) that should be used for delaying
+     * a network call response.
+     */
+    private static int calculateDelayForCall() {
+        // Commenting out the random failure mode delay since we want our
+        // tests to be reproducible
+        return //calculateIsFailure() ? calculateDelayForError() :
+                calculateDelayForSuccess();
+    }
+
+    /**
+     * Match url to a regex template while allowing extra slash and query
+     * strings at the end
+     */
+    private static boolean urlMatches(String url, String template) {
+        if (TextUtils.isEmpty(url) || TextUtils.isEmpty(template)) {
+            return false;
+        }
+        String pattern = '^' + template;
+        if (template.charAt(template.length() - 1) != '/') {
+            pattern += "/?";
+        }
+        pattern += "(\\?.*)?$";
+        return url.matches(pattern);
+    }
+
+    private MockResponse generateMockResponse(RecordedRequest request) {
+        final String method = request.getMethod();
+        final String path = request.getPath();
+        MockResponse response = new MockResponse();
+        response.addHeader("Set-Cookie", "csrftoken=dummy; Max-Age=31449600; Path=/");
+        response.setResponseCode(404);
+        if ("POST".equals(method)) {
+            if (urlMatches(path, "/oauth2/access_token")) {
```

Comment:
Seems like there must be a nicer way to do this. Make a fake response type with an HTTPMethod, a URL matcher, and response data, then people can just add items to that list, and the dispatch code gets a lot smaller since it's just iterating over a list.